crates/wysiwyg/src/dom/parser/parse.rs

```diff
@@ -0,0 +1,200 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::dom::parser::PaNodeContainer;
+use crate::dom::{ContainerNode, Dom, DomCreationError, DomNode, TextNode};
+use crate::ToHtml;
+
+use super::padom_node::PaDomNode;
+use super::{PaDom, PaDomCreationError, PaDomCreator};
+
+pub fn parse(html: &str) -> Result<Dom<u16>, DomCreationError<u16>> {
+    PaDomCreator::parse(html)
+        .map(padom_to_dom_u16)
+        .map_err(padom_creation_error_to_dom_creation_error)
+}
+
+/// Convert a [PaDom] into a [Dom].
+///
+/// [PaDom] is purely used within the parsing process (using html5ever) - in it,
+/// parents refer to their children by handles, and all the nodes are owned in
+/// a big list held by the PaDom itself. PaDoms may also contain garbage nodes
+/// that were created during parsing but are no longer needed. A garbage
+/// collection method was written for testing and is inside padom_creator's
+/// test code. The conversion process here ignores garbage nodes, so they do
+/// not appear in the final Dom.
+///
+/// [Dom] is for general use. Parent nodes own their children, and Dom may be
+/// cloned, compared, and converted into an HTML string.
+fn padom_to_dom_u16(padom: PaDom) -> Dom<u16> {
+    /// Recurse into panode's children and convert them too
+    fn convert_children(
+        padom: &PaDom,
+        child: &PaNodeContainer,
+        new_node: Option<&mut DomNode<u16>>,
+    ) {
+        if let DomNode::Container(new_node) = new_node.unwrap() {
+            convert(padom, child, new_node);
+        } else {
+            panic!("Container became non-container!");
+        }
+    }
+
+    /// Create a formatting node
+    fn new_formatting(tag: &str) -> DomNode<u16> {
+        DomNode::Container(ContainerNode::new_formatting(
+            tag.to_html(),
+            Vec::new(),
+        ))
+    }
+
+    /// Create a link node
+    fn new_link(child: &PaNodeContainer) -> DomNode<u16> {
+        DomNode::Container(ContainerNode::new_link(
+            child.get_attr("href").unwrap_or("").to_html(),
+            Vec::new(),
+        ))
+    }
+
+    /// Copy all panode's information into node (now we know it's a container).
+    fn convert_container(
+        padom: &PaDom,
+        child: &PaNodeContainer,
+        node: &mut ContainerNode<u16>,
+    ) {
+        let tag = child.name.local.as_ref();
+        match tag {
+            "b" | "i" | "strong" | "em" | "del" | "u" => {
```

Comment:
We should probably add "code" to the list of formatting tags.

Reply:
Good catch, fixed.